scripts/spack_manifest/injection/projections.py

```
@@ -0,0 +1,236 @@
+    if not root_spec_definition:
+        raise ValueError(
+            f"Could not find root spec definition for projection {projection} in the manifest. The spack.yaml is invalid"
+        )
+
+    # Now we can extract the version from the root spec definition
+    version_regex = re.compile(rf"{projection}@(?:git.)?([^+~= ]+).*")
+    match = re.match(version_regex, root_spec_definition)
+
+    if not match:
+        raise ValueError(
+            f"Could not extract version from root spec definition {root_spec_definition} for projection {projection}. The root spec needs a version defined."
+        )
+
+    version = match.group(1)
+
+    print(
+        f"Extracted version {version} from root spec definition {root_spec_definition}"
+    )
+
+    return version
+
+
+def inject_projections(
+    manifest_path: str, root_spec: str, packages: set[str]
+) -> dict[str, Any]:
+
+    with open(manifest_path, "r") as file:
+        manifest: dict[str, Any] = yaml.safe_load(file)
+
+    # We should try and inject projections for the root spec as well, so we create a set with just it
+    root_spec_set: set[str] = {root_spec}
+
+    # Get projections that are already defined in the manifest - we don't want to redefine these
+    defined_projections: dict[str, str] = _get_defined_projections(manifest)
+    defined_projections_set: set[str] = set(defined_projections.keys())
+    # Get packages that have versions defined in the manifest - we can only generate projections for things with an explicit version
+    packages_with_versions_defined: set[str] = _get_packages_with_versions_defined(
+        manifest
+    )
+
+    # Essentially...we want to generate projections for all packages that don't already have them, provided they have a version defined, or they are the root spec.
+    projections_to_generate: set[str] = (
+        root_spec_set | (packages & packages_with_versions_defined)
+    ) - defined_projections_set
```

This is really elegant. I like it. 

However .. if I pass nonsense/missing packages on the command line they are silently ignored.

Could we check if there are packages that are defined but not present in the manifest and warn a user? I guess it's unlikely in the CI workflow, but might happen if this module is used in other contexts.

---

Could it be valid to have a package defined in `--packages` but have no entry in `spack.packages.*`? Like, we want to keep track of this package but don't have any requirements to place on it?

---

Missing packages would be a bit difficult to find since they could be a dependency that isn't explicitly specified in the abstract manifest, but only show in the concretized version

---

> Could it be valid to have a package defined in --packages but have no entry in spack.packages.*? Like, we want to keep track of this package but don't have any requirements/constraints to place on it?

Just for completeness, we discussed IRL and decided we would not support generating packages that did not have a version specified in the packages section. This has been implemented I believe, so will resolve this.